stackslib/src/net/download/nakamoto.rs

```diff
@@ -0,0 +1,4283 @@
+                rc,
+                pox_constants
+                    .block_height_to_reward_cycle(first_burn_height, wt_start.burn_height)
+                    .expect("FATAL: tenure from before system start"),
+                wt.processed,
+            );
+            tenure_start_end.fetch_end_block = true;
+            test_debug!(
+                "i={}, k={}, n={}, len={}, next_len={}; {:?}",
+                i,
+                k,
+                n,
+                wanted_tenures.len(),
+                next_wanted_tenures.len(),
+                &tenure_start_end
+            );
+            tenure_block_ids.insert(wt.tenure_id_consensus_hash.clone(), tenure_start_end);
+        }
+        Some(tenure_block_ids)
+    }
+}
+
+/// A set of confirmed downloader state machines assigned to one or more neighbors.  The block
+/// downloader runs tenure-downloaders in parallel, since the downloader for the N+1'st tenure
+/// needs to feed data into the Nth tenure.  This struct is responsible for scheduling peer
+/// connections to downloader state machines, such that each peer is assigned to at most one
+/// downloader.  A peer is assigned a downloader for the duration of at most one RPC request, at
+/// which point, it will be re-assigned a (possibly different) downloader.  As such, each machine
+/// can make progress even if there is only one available peer (in which case, that peer will get
+/// scheduled across multiple machines to drive their progress in the right sequence such that
+/// tenures will be incrementally fetched and yielded by the p2p state machine to the relayer).
+pub struct NakamotoTenureDownloaderSet {
+    /// A list of instantiated downloaders that are in progress
+    pub(crate) downloaders: Vec<Option<NakamotoTenureDownloader>>,
+    /// An assignment of peers to downloader machines in the `downloaders` list.
+    pub(crate) peers: HashMap<NeighborAddress, usize>,
+    /// The set of tenures that have been successfully downloaded (but possibly not yet stored or
+    /// processed)
+    pub(crate) completed_tenures: HashSet<ConsensusHash>,
+}
+
+impl NakamotoTenureDownloaderSet {
+    pub fn new() -> Self {
+        Self {
+            downloaders: vec![],
+            peers: HashMap::new(),
+            completed_tenures: HashSet::new(),
+        }
+    }
+
+    /// Assign the given peer to the given downloader state machine.  Allocate a slot for it if
+    /// needed.
+    fn add_downloader(&mut self, naddr: NeighborAddress, downloader: NakamotoTenureDownloader) {
+        test_debug!(
+            "Add downloader for tenure {} driven by {}",
+            &downloader.tenure_id_consensus_hash,
+            &naddr
+        );
+        if let Some(idx) = self.peers.get(&naddr) {
+            self.downloaders[*idx] = Some(downloader);
+        } else {
+            self.downloaders.push(Some(downloader));
+            self.peers.insert(naddr, self.downloaders.len() - 1);
```

Comment:
Is it every possible that self.downloaders is empty? I am assuming not, and it if was though would we want to error? just worried this might become a nebulous underflow that causes harder to track down errors.